### DISCUSSION TITLE
Arguments for moving Interactive Examples into mdn/content + mdn/yari

### DISCUSSION THREAD

**Discussion by peterbe** [mdn/mdn-community#183] [Content]
At the moment, we a completely separate infrastructure for producing URLs like https://interactive-examples.mdn.mozilla.net/pages/tabbed/video.html
The content is in https://github.com/mdn/interactive-examples and https://github.com/mdn/bob is used to compile the content into web pages. 

I think it makes a lot of sense to merge all of this into mdn/content and mdn/yari. But it's possible to do only parts of that too. E.g. leave the content in github.com/mdn/interactive-examples but use mdn/yari instead of mdn/bob. 

### Pros

- mdn/content has "community" and stuff like write-access members and `CODEOWNERS` all figured out. In one place. 
- monorepos's inherent advantages (e.g. all text content and examples all in one place which can make it easy and pleasant to make a PR that touches both)
- can remove all NPM publishing infrastructure in bob that is poorly maintained
- Fewer Devops things to maintain
- We can better control problems in the use of `{{InteractiveExamples('something-twhats-a-typo')}}` to avoid things like https://github.com/mdn/content/issues/3784) rather than hoping it will work
- Both Yari and Interactive examples lack CSP and if it's all under one roof it'll be easier to make secure because fewer things to manage.

### Cons

- impurity in that mdn/content ceases to be exclusively about one thing (the `index.{html|md}` files)
- monorepos' inherent disadvantages (e.g. if you file an issue you have to prefix that you're about to talk about an interactive example)
- interactive examples rarely change but if included in Yari they'd have to be rebuilt every 24h

**Comment by peterbe** [comment]
mdn/interactive-examples is still using TravisCI. In mdn/content we have all the GitHub Actions stuff figured out :)

**Reply by Ryuno-Ki** [reply]
Picking up on https://github.com/mdn/content/issues/4108 (and the issues linked from there) I have to admit, I got pretty confused.

I like our interactive demos, but I feel, there are different types of them inlined as iframe into the HTML.
Which kind of makes sense, since you can sandbox them this way instead of running the risk that they would dabble with the HTML around it.
Plus, you have a place to apply linters, tests etc.

So from a security and DX point of view, I'd vote for keeping the examples separate from the content.

The macro argument is similar to other data sources we use (like BCD). Perhaps a similar API could be queried?
(Like, GitHub API?)

**Reply by peterbe** [reply]
> So from a security and DX point of view, I'd vote for keeping the examples separate from the content.

My discussion here is about whether it should be a separate **repo** or not. 
Because interactive examples are potentially repeated on multiple pages (including the translations) it's smart that they're not tied to a specific page, like the live samples are. Sure. But the question is, are there strong enough benefits to lifting the interactive examples out of mdn/interactive-example and plopping them into mdn/content (separate from the regular `index.html` files)? ...and in the process, fold mdn/bob into mdn/yari instead.

**Reply by Ryuno-Ki** [reply]
If they would be included in `mdn/content`, would they still be deployed as a different origin in the final page?

**Reply by peterbe** [reply]
Yeah, the end result wouldn't change. It'd still be a dedicated different domain name. 
But it'll be more similar to live samples.

**Reply by Ryuno-Ki** [reply]
Sounds like worth a PoC then.

**Comment by Ryuno-Ki** [comment]
First thing that spring into my mind:

![Planet Bob of Titan AE](http://images6.fanpop.com/image/photos/36700000/Titan-AE-image-titan-ae-36734764-500-180.gif)

Will give constructive feedback in the thread :)

**Comment by wbamberg** [comment]
I think this would be fine in principle. interactive-examples is a separate repo because when it was invented the content was in the Wiki. If the content had been in GitHub back then, interactive-examples would probably have gone in the same repo.

I think there are definitely advantages to contributors to be able to file one PR to add an interactive example and update the page at the same time. It's a bit hokey [the way it is now](https://github.com/mdn/interactive-examples/pull/1799#issuecomment-829561015).

At the moment you can do:

```
npm install
npm run start
```

and get a local server that serves the examples. This is nice! Would we get something similar?

How would we organize the files? Is the suggestion basically to have "interactive-examples" as a top level directory, alongside "files"?

What about other examples repos, notably https://github.com/mdn/css-examples and https://github.com/mdn/dom-examples (but there are others)? These are usually (but perhaps not always) designed to be run as github pages, and sometimes embedded in MDN pages using the [EmbedGHLiveSample](https://github.com/mdn/yari/blob/main/kumascript/macros/EmbedGHLiveSample.ejs) macro. Should we instead have:

```
/examples
  /interactive-examples
  /css-examples
  /dom-examples
  /...
```

?

Some day we should rationalize our examples system(s). I appreciate we don't want to do that as part of this proposal, but it would be good to try to make this a move that's at least compatible with a bigger plan.

@Rumyra might be interested in this discussion.

**Reply by peterbe** [reply]
Quick question; would it be an option to put each interactive example into the page that displays it?
E.g. `files/en-us/web/foo/bar/index.html` and `files/en-us/web/foo/bar/interactive-example/example1/index.(css|js|html)`

There are 968 instances of `{{EmbedInteractiveExample(...)}}` and only 17 of those are "repeated". E.g. `"pages/js/array-push.html"` is used in 3 different `index.html` files. 
Perhaps the argument can be implied by the location. E.g. `./interactive-example/**` or it could be in the front-matter. 
Then in 951 of the pages you can just say `{{EmbedInteractiveExample}}`
And if you need to refer to an interactive example from a different location, you can do `{{EmbedInteractiveExample("Web/That/Other/Page")}}`

**Reply by peterbe** [reply]
By the way, not a single page, that isn't mdn/kitchensink or mdn/structures/code_examples uses more than 1 interactive example. 

If we embed them within the content, and just use run-time Yari to preview, you don't need any additional tooling (e.g. `npm run start`). It would just be a matter of editing `files/en-us/web/foo/bar/examples/interactive-examples/example1/index.js` and reloading <http://localhost:3000/en-us/docs/web/foo/bar>

**Reply by wbamberg** [reply]
I think this would be fine, again in principle. I expect that tying the examples more closely to their pages would be a good thing overall. As you say, we've had these things a long time and haven't seen a lot of examples used in more than one page.

Lots of interactive example share things between themselves though, like common CSS or media assets, or fonts. So it would be nice to have a way to address that without it being a rats' nest.

> By the way, not a single page, that isn't mdn/kitchensink or mdn/structures/code_examples uses more than 1 interactive example.

It at least used to be the case, in Kuma, that having >1 in a page broke stuff. But anyway yes, the current model for interactive examples is very specific: there's one per reference page, at the top of the page.

> If we embed them within the content, and just use run-time Yari to preview, you don't need any additional tooling (e.g. npm run start). It would just be a matter of editing files/en-us/web/foo/bar/examples/interactive-examples/example1/index.js and reloading http://localhost:3000/en-us/docs/web/foo/bar

Yes please!

**Reply by peterbe** [reply]
Personally, I'd need to familiarize myself with more details to fully understand the implications of some interactive examples sharing assets like CSS. But I quite like the idea of co-locating stuff as much as possible.

**Reply by Rumyra** [reply]
I am on board with moving interactive examples _into_ content or yari. My pref would be content with co-location. It would certainly make the workflow of adding/updating interactive examples easier.

Apologies if I've missed some discussion about this, but would they still be embedded in an iframe, or run within the page?

**Reply by peterbe** [reply]
> would they still be embedded in an iframe, or run within the page?

Yes. 
There might a future where we stop needing an iframe because we've security-vetted all interactive examples. 

But still, a way to break this up is to just rearrange how content (no pun intended) is organized. We'd probably need to rewrite `bob` which is the code that packages the interactive example into an HTML page that would be served inside the iframe. 
If we do that, apart from minor differences in the `bob -> Yari builder` the end-user wouldn't notice any differences. 

Heck, it might be easiest to first just put the `bob` code into mdn/yari so as to not break what's working.

**Comment by pepelsbey** [comment]
We’re currently in the process of moving examples from GH pages into content.

**Reply by wbamberg** [reply]
This isn't about that. This is a proposal to move _interactive examples_ into content, not GH examples.
